workspaces/cli-shared/src/diffs/trail-parsers.ts

```diff
@@ -74,84 +62,105 @@ export function locationForTrails(
         ? trail.SpecRequestRoot
         : trail.SpecRequestBody;
 
-    let request = null;
-    for (const endpoint of endpoints) {
-      for (const requestBody of endpoint.requestBodies) {
-        if (requestBody.requestId === requestId) {
-          request = requestBody;
-          break;
-        }
-      }
-      if (request) {
-        break;
-      }
-    }
+    let contentType = '';
+    const endpoint = endpoints.find(
+      (endpoint) =>
+        !!endpoint.requestBodies.find((requestBody) => {
+          if (requestBody.requestId === requestId) {
+            contentType = requestBody.contentType;
+          }
+          return requestBody.requestId === requestId;
+        })
+    );
 
-    if (!request) {
+    if (!endpoint) {
       console.error(
         `Could not find endpoint with request ${requestId} in current spec endpoints`
       );
       return null;
     }
 
-    const { method, pathId, contentType } = request;
+    const { method, pathId } = endpoint;
 
     return {
       pathId: pathId,
       method: method,
-      requestId,
-      contentType,
-      inRequest: true,
+      descriptor: {
+        type: 'request',
+        requestId,
+        contentType,
+      },
     };
   } else if ('SpecResponseRoot' in trail || 'SpecResponseBody' in trail) {
     const { responseId } =
       'SpecResponseRoot' in trail
         ? trail.SpecResponseRoot
         : trail.SpecResponseBody;
-    let response = null;
-    for (const endpoint of endpoints) {
-      for (const responseBody of endpoint.responseBodies) {
-        if (responseBody.responseId === responseId) {
-          response = responseBody;
-          break;
-        }
-      }
-      if (response) {
-        break;
-      }
-    }
 
-    if (!response) {
+    let contentType = '';
+    let statusCode = 0;
+    const endpoint = endpoints.find(
+      (endpoint) =>
+        !!endpoint.responseBodies.find((responseBody) => {
+          if (responseBody.responseId === responseId) {
+            contentType = responseBody.contentType;
+            statusCode = responseBody.statusCode;
+          }
+          return responseBody.responseId === responseId;
+        })
+    );
+
+    if (!endpoint) {
       console.error(
         `Could not find endpoint with response ${responseId} in current spec endpoints`
       );
       return null;
     }
-    const { pathId, method, statusCode, contentType } = response;
+    const { pathId, method } = endpoint;
 
     return {
       pathId: pathId,
       method: method,
-      statusCode: statusCode,
-      responseId,
-      contentType,
-      inResponse: true,
+      descriptor: {
+        type: 'response',
+        statusCode: statusCode,
+        responseId,
+        contentType,
+      },
     };
   } else if ('SpecPath' in trail) {
```

Comment:
Do you know when this trail is generated? It doesn't have a request / response id, but mimics the shape of them and later on we coerce these to non-nullable (so it appears to not crash the UI currently) - but I'm really confused what this is supposed to represent

Reply:
Ok it's generated from somewhere, found out from the snapshot tests - we should make these explicit, but I think this would do with a good refactor to split up diff types such we can define an abstract class, and depending on the diff type we have specific classes that'll generate spec trails + descriptions, etc).

Reply:
These trails are generated by the engine, as a result from the diffing, here: https://github.com/opticdev/optic/blob/ce222d9dd130b6cfd11b8415a5a1a03733e85c5b/workspaces/optic-engine/src/interactions/result.rs#L627-L635.


For a quick overview, there's 4 types of trails you'll find returned by the engine:

- [interaction trail](https://github.com/opticdev/optic/blob/ce222d9dd130b6cfd11b8415a5a1a03733e85c5b/workspaces/optic-engine/src/interactions/result.rs#L493-L495): points to a certain place inside a _specific_ interaction. Examples: it's query parameters, the request body, a response with a certain status code, etc.
- [requests trail / request spec trail](https://github.com/opticdev/optic/blob/ce222d9dd130b6cfd11b8415a5a1a03733e85c5b/workspaces/optic-engine/src/interactions/result.rs#L627-L635) (should actually be called `EndpointSpecTrail` probably): pointing to a place in the spec by their ids on an endpoint level. Examples: a path by id,  a request body by id and content type, query parameters by a query parameters id, etc. They are mostly used in diffs to indicate the interaction matched against this particular part of the spec.
- [json trails](https://github.com/opticdev/optic/blob/ce222d9dd130b6cfd11b8415a5a1a03733e85c5b/workspaces/optic-engine/src/shapes/traverser.rs#L305-L307): point to a location inside a JSON body, most likely coming from an interaction. Example: the third array item nested inside the "author" field of an object. As we're moving of only JSON as the source, this is probably better named "body trail".
- [shape trail](https://github.com/opticdev/optic/blob/ce222d9dd130b6cfd11b8415a5a1a03733e85c5b/workspaces/optic-engine/src/shapes/traverser.rs#L273-L276): points to a location inside the shape part of the spec. Examples: the shape id of an array, the parent shape id of an object field, etc. Like the request spec trails, they are mostly used in diffs (shape diffs in this case) to indicate something matched against the spec.

Hopefully that's useful in some way 😅.